### ISSUE TITLE
Performance of the via-IR pipeline

### ISSUE THREAD

**Issue by NunoFilipeSantos** [ethereum/solidity#13719]
### What

The Yul optimizer is slow.

### Why

Parts have quadratic runtime complexity.

### How

Identify the responsible components and reimplement them more efficiently.

Low-hanging fruits in improving the stack shuffling algorithm (replace per-iteration-regenerated multiplicity maps).

#### Tasks

- [x] https://github.com/ethereum/solidity/pull/13705
- [x] https://github.com/ethereum/solidity/pull/13703
- [x] https://github.com/ethereum/solidity/pull/13734
- [x] https://github.com/ethereum/solidity/pull/13760
- [x] https://github.com/ethereum/solidity/pull/13702
- [x] https://github.com/ethereum/solidity/pull/13708
- [x] https://github.com/ethereum/solidity/pull/13688
- [x] https://github.com/ethereum/solidity/pull/13687
- [x] https://github.com/ethereum/solidity/pull/13682
- [x] https://github.com/ethereum/solidity/pull/13752
- [x] https://github.com/ethereum/solidity/pull/12274
- [x] https://github.com/ethereum/solidity/issues/13822

# Paused
- [ ] https://github.com/ethereum/solidity/pull/14112
- [ ] Community Demo or Blog post

### Out of scope

- At all cost beat the legacy pipeline in runtime.

---
### Resources

1. [Performance Statistics](https://github.com/ethereum/solidity/issues/13719#issuecomment-1324133162)
2. https://github.com/ethereum/solidity/issues/13721

**Comment by nikola-matic** [comment]
@chriseth would it make sense to cherry pick #13703 into the most recent develop without the optimiser changes that you've introduced in order to have a control/baseline for the optimiser performance prior to actually introducing any patches?

**Comment by chriseth** [comment]
Performance statistics:
==================

(note that some changes were merged in between that do not only change the performance but also the optimizer steps themselves, so take these statistics with a grain of salt)

Run on "Chains.sol" on @chriseth's machine with RelWithDebInfo `--via-ir --bin --optimize`.

--------

Run on 9a60c06335bb49582a36b17918f2114c44165951 (before https://github.com/ethereum/solidity/pull/13688 was merged)
```

  0.669% (0.207962 s): FullInliner
  1.087% (0.337953 s): UnusedPruner
  1.193% (0.370938 s): EqualStoreEliminator
  1.530% (0.47585 s): SSATransform
  2.143% (0.666361 s): UnusedStoreEliminator
  5.612% (1.745471 s): ExpressionSimplifier
  8.146% (2.533345 s): LoadResolver
 12.907% (4.014192 s): LiteralRematerialiser
 23.050% (7.168812 s): UnusedAssignEliminator
 41.923% (13.038435 s): CommonSubexpressionEliminator
--------------------------------------
    100% (31.101 s)

  1.163% (0.148087 s): SSATransform
  1.275% (0.162394 s): UnusedPruner
  1.528% (0.194537 s): UnusedStoreEliminator
  2.109% (0.268484 s): EqualStoreEliminator
  4.215% (0.536716 s): ExpressionSimplifier
  8.557% (1.089507 s): LoadResolver
 10.258% (1.306054 s): LiteralRematerialiser
 28.750% (3.66045 s): UnusedAssignEliminator
 39.864% (5.075496 s): CommonSubexpressionEliminator
--------------------------------------
    100% (12.732 s)

56.7s

```

-----

Run on 19fc395662c84994f9c4094afd60b83ca69fdf13 (after #13688 was merged)
```
  0.663% (0.196783 s): FullInliner
  1.127% (0.334353 s): UnusedPruner
  1.373% (0.407598 s): EqualStoreEliminator
  1.602% (0.475363 s): SSATransform
  2.005% (0.595169 s): UnusedStoreEliminator
  5.976% (1.77352 s): ExpressionSimplifier
  7.799% (2.31465 s): LoadResolver
 13.117% (3.893043 s): LiteralRematerialiser
 24.148% (7.166617 s): UnusedAssignEliminator
 40.453% (12.005757 s): CommonSubexpressionEliminator
--------------------------------------
    100% (29.678 s)

  1.235% (0.14263 s): UnusedPruner
  1.254% (0.144803 s): SSATransform
  1.550% (0.178966 s): UnusedStoreEliminator
  2.328% (0.268853 s): EqualStoreEliminator
  4.734% (0.546637 s): ExpressionSimplifier
  8.352% (0.964534 s): LoadResolver
 10.361% (1.196546 s): LiteralRematerialiser
 29.840% (3.445972 s): UnusedAssignEliminator
 38.108% (4.400843 s): CommonSubexpressionEliminator
--------------------------------------
    100% (11.548 s)

54.0s
```

----

Run on 1bd35b1be (after #13682 - use hash for CSE was merged)
```
  0.350% (0.077108 s): DeadCodeEliminator
  0.892% (0.196741 s): FullInliner
  1.451% (0.320119 s): UnusedPruner
  1.650% (0.364055 s): EqualStoreEliminator
  2.129% (0.469621 s): SSATransform
  2.777% (0.612714 s): UnusedStoreEliminator
  7.836% (1.728824 s): ExpressionSimplifier
 10.842% (2.391972 s): LoadResolver
 17.522% (3.865645 s): LiteralRematerialiser
 21.876% (4.826155 s): CommonSubexpressionEliminator
 30.729% (6.779252 s): UnusedAssignEliminator
--------------------------------------
    100% (22.061 s)

  1.601% (0.144756 s): SSATransform
  1.642% (0.148393 s): UnusedPruner
  1.992% (0.180077 s): UnusedStoreEliminator
  2.712% (0.245104 s): EqualStoreEliminator
  5.861% (0.529793 s): ExpressionSimplifier
 11.542% (1.043309 s): LoadResolver
 13.444% (1.215262 s): LiteralRematerialiser
 21.098% (1.907142 s): CommonSubexpressionEliminator
 37.191% (3.361847 s): UnusedAssignEliminator
--------------------------------------
    100% (9.039 s)

44.6s
```

----
Run on be8ecb17d8e8405 (develop as of 2022-11-22 evening)

```
  0.885% (0.203343 s): FullInliner
  1.530% (0.351829 s): UnusedPruner
  1.616% (0.371587 s): EqualStoreEliminator
  2.070% (0.47594 s): SSATransform
  2.728% (0.627155 s): UnusedStoreEliminator
  7.599% (1.746962 s): ExpressionSimplifier
 10.675% (2.453971 s): LoadResolver
 17.620% (4.050501 s): LiteralRematerialiser
 21.554% (4.954944 s): CommonSubexpressionEliminator
 31.397% (7.217784 s): UnusedAssignEliminator
--------------------------------------
    100% (22.988 s)

  1.522% (0.149838 s): UnusedPruner
  1.635% (0.16091 s): SSATransform
  2.058% (0.202598 s): UnusedStoreEliminator
  2.954% (0.290798 s): EqualStoreEliminator
  6.055% (0.596098 s): ExpressionSimplifier
 10.950% (1.077955 s): LoadResolver
 12.650% (1.245365 s): LiteralRematerialiser
 20.915% (2.059034 s): CommonSubexpressionEliminator
 38.458% (3.786018 s): UnusedAssignEliminator
--------------------------------------
    100% (9.845 s)

46.0s
```

----
Run after https://github.com/ethereum/solidity/pull/13705 was merged:

```
  0.911% (0.190328 s): FullInliner
  1.370% (0.286232 s): UnusedPruner
  1.694% (0.353885 s): EqualStoreEliminator
  2.234% (0.466618 s): SSATransform
  2.675% (0.558851 s): UnusedStoreEliminator
  8.033% (1.678058 s): ExpressionSimplifier
 10.560% (2.205888 s): LoadResolver
 18.033% (3.766786 s): LiteralRematerialiser
 22.306% (4.659482 s): CommonSubexpressionEliminator
 29.865% (6.238317 s): UnusedAssignEliminator
--------------------------------------
    100% (20.889 s)

  0.195% (0.016969 s): FullInliner
  0.196% (0.017102 s): LoopInvariantCodeMotion
  0.213% (0.018522 s): ConditionalUnsimplifier
  0.238% (0.020727 s): ControlFlowSimplifier
  0.241% (0.021015 s): ConditionalSimplifier
  0.259% (0.022609 s): DeadCodeEliminator
  0.452% (0.039353 s): Rematerialiser
  0.572% (0.049845 s): ExpressionSplitter
  1.475% (0.12856 s): UnusedPruner
  1.567% (0.136607 s): SSATransform
  2.150% (0.187418 s): UnusedStoreEliminator
  3.064% (0.267053 s): EqualStoreEliminator
  6.249% (0.544632 s): ExpressionSimplifier
 11.037% (0.962025 s): LoadResolver
 13.590% (1.184535 s): LiteralRematerialiser
 21.429% (1.867744 s): CommonSubexpressionEliminator
 36.591% (3.189325 s): UnusedAssignEliminator
--------------------------------------
    100% (8.716 s)

42.7 s
```

**Comment by NunoFilipeSantos** [comment]
The first scope was complete.
